File: Solutions/Python/zigzag_conversion.py
"""
    [MEDIUM]
    6. Zigzag Conversion

    Concepts:
    - string

    Stats:
        Runtime | 13 ms     [Beats 45.40%]
        Memory  | 17.82 MB  [Beats 41.03%]
"""

class Solution:
    def convert(self, s: str, num_rows: int) -> str:
        str_dirs = []
        is_moving_down = True
        k = 0

        for i, c in enumerate(s):
            str_dirs.append((c, k))
            k += 1 if is_moving_down else -1

            if k == 0:
                is_moving_down = True
            elif k == num_rows - 1:
                is_moving_down = False

        str_dirs.sort(key=lambda x: x[1])
        return "".join([c for c, _ in str_dirs])

File: Solutions/Rust/zigzag_conversion.rs
/*
    [MEDIUM]
    6. Zigzag Conversion

    Concepts:
    - string

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.32 MB   [Beats 39.26%]
*/
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let mut str_dirs: Vec<(char, i32)> = Vec::new();
        let mut is_moving_down: bool = true;
        let mut k: i32 = 0;

        for (i, c) in s.chars().enumerate() {
            str_dirs.push((c, k));
            k += if is_moving_down { 1 } else { -1 };

            if k == 0 {
                is_moving_down = true;
            } else if k == num_rows - 1 {
                is_moving_down = false;
            }
        }

        str_dirs.sort_by_key(|&item| item.1);
        return str_dirs
            .into_iter()
            .map(|item| item.0)
            .collect();
    }
}